### PR TITLE
Add max_values and argmax convenience functions to ATen

### LINKED ISSUE #5616: [feature request] torch.argmax / torch.argmin
To save indexing into output of torch.max / torch.min and make code more readable. NumPy / TensorFlow also has it.

### PR DESCRIPTION
**Pull request by goldsborough**
This PR adds convenience functions to only get the min/max values, or only the min/max indices, of a tensor. This avoids having to write `std::get<0>(tensor.max(dim))`.

Part of #4619 (CC @ebetica)

Not sure if we want to expose these in Python as well? `max(dim)[0]` is more convenient in python. Although `torch.argmax()` is also nice to have, since numpy has it too. I can add documentation to the functions if we want to make them part of the Python API.

closes #5616 

@apaszke @colesbury @ezyang

### CODE REVIEW

**Comment by zou3519** [comment]
I think torch.argmax / torch.argmin would be good to have in Python. Users have been requesting the feature: https://github.com/pytorch/pytorch/issues/5616

**Comment by goldsborough** [comment]
@zou3519 You mean defaulting `dim=0`?

**Comment by colesbury** [comment]
You'll have to add it to SKIP_PYTHON_BINDINGS in `tools/autograd/gen_python_functions.py` if you don't want it exposed to Python.

I also think `argmax` and `argmin` would be reasonable in Python, but deserve docs.

FWIW, for `max`/`min`, I think we will eventually want to unify the overloads Python and add support for multiple dimensions. Perhaps something like:

```
max(input, dim=None, return_indices=None) -> Tensor or (Tensor, Tensor)
```

Where `dim` can be `None`, an `int`, or a tuple of `int`s and `return_indices` controls whether the function returns indices. (With `return_indices` being the legacy behavior dependent on `dim`).

**Comment by goldsborough** [comment]
Ok, so I think what I'll do then is:
1. Make `argmin`/`argmax` a proper part of the Python API with documentation, so we can include it in the next release,
2. Use `SKIP_PYTHON_BINDINGS` for `min_values`/`max_values` and,
3. Create an issue to implement the unified min/max function.

**Comment by ezyang** [comment]
@colesbury I CC'ed you in another issue, but what to raise it again here: what about `torch.max(x, y)`?

**Comment by ezyang** [comment]
Needs a test

**Comment by zou3519** [comment]
@goldsborough I meant defaulting `dim=None`, so `argmax(tensor)` gives the index of the largest element in the tensor without a dim reduction.

**Comment by colesbury** [comment]
We should follow the NumPy convention of default to `argmax` over the entire array. This matches other reductions, which operate over the entire array if a dim isn't specified.

**Comment by goldsborough** [comment]
Ok so I've implementd `torch.argmin()` and `torch.argmax()` such that `torch.argmax(x, dim=None)` is equivalent to `torch.argmax(x.view(-1), dim=0, keepdim=False)`. For this I had to name the native function exposed from C++ `_argmax` (with underscore) and not expose `argmax`.

I also added an overload in ATen/C++ that mimics the behavior of `dim=None`.

File: aten/src/ATen/native/TensorCompare.cpp [hunk 1]
```diff
@@ -79,4 +79,43 @@ Tensor _s_where_cpu(const Tensor& condition, const Tensor& self, const Tensor& o
...
+}
+
+// argmax and argmin
+
+Tensor argmax(const Tensor& self, int64_t dim, bool keepdim) {
+  return std::get<1>(self.max(dim, keepdim));
+}
+
+Tensor argmax(const Tensor& self) {
+  return std::get<1>(self.contiguous().view({-1}).max(/*dim=*/0));
```
Comment (fmassa) [inline aten/src/ATen/native/TensorCompare.cpp]: nit: you could also use `self.reshape({-1})` I think

Comment (goldsborough) [reply aten/src/ATen/native/TensorCompare.cpp]: Oh ok. Does it ensure the tensor is contiguous?

**Comment by goldsborough** [comment]
I think this is ready? @apaszke @colesbury

**Review by apaszke — APPROVED** [review]

**Comment by apaszke** [comment]
LGTM. I only don't understand why we need to expose them with an underscore and wrap again in Python. I'm a bit worried about the implications for the JIT.

**Comment by goldsborough** [comment]
Because we wanted to allow `dim=None` in Python and route to a different function. Essentially the problem was wanting to wrap the C++ function in Python. AFAIK there is no way of translating `dim=None` behavior into C++? So Richard and I came up with this strategy

**Comment by fmassa** [comment]
Maybe `at::optional` could be used?

**Comment by goldsborough** [comment]
yeah but we'd have to make a lot of changes to the codegen I believe?

I notice I forgot to commit the `reshape({-1})` change. I have it locally and forgot to push 😢
